File: python-example/streaming_ingest_example.py
"""
Example demonstrating how to use the Snowflake Streaming Ingest SDK in Python
with the high-performance architecture and default pipe.

The default pipe is automatically created by Snowflake when you first
open a channel. No CREATE PIPE DDL is required. The default pipe name
follows the convention: <TABLE_NAME>-STREAMING
"""

from datetime import datetime
import uuid
import os

# Change Environment Variable SS_LOG_LEVEL="info" to increase logging details
os.environ["SS_LOG_LEVEL"] = "warn"

from snowflake.ingest.streaming import StreamingIngestClient


MAX_ROWS = 100_000

# Replace these with your Snowflake object names
DATABASE = "MY_DATABASE"
SCHEMA = "MY_SCHEMA"
TABLE = "MY_TABLE"

# Default pipe: Snowflake auto-creates this on first channel open
PIPE = f"{TABLE}-STREAMING"


def main():
    """Main function to demonstrate streaming data ingestion."""

    # Create Snowflake Streaming Ingest Client using context manager
    with StreamingIngestClient(
        client_name=f"MY_CLIENT_{uuid.uuid4()}",
        db_name=DATABASE,
        schema_name=SCHEMA,
        pipe_name=PIPE,
        profile_json="profile.json"
    ) as client:

        print("Client created successfully")

        # Open a channel for data ingestion using context manager
        with client.open_channel(f"MY_CHANNEL_{uuid.uuid4()}")[0] as channel:
            print(f"Channel opened: {channel.channel_name}")

            # Ingest rows — column names must match the target table schema.
            # The default pipe uses MATCH_BY_COLUMN_NAME to map fields.
            print(f"Ingesting {MAX_ROWS} rows...")
            for i in range(1, MAX_ROWS + 1):
                row_id = str(i)
                channel.append_row(
                    {
                        "c1": i,
                        "c2": row_id,
                        "ts": datetime.now()
                    },
                    row_id
                )

                if i % 10_000 == 0:
                    print(f"Ingested {i} rows...")

            print("All rows submitted. Waiting for commit...")

            # Wait for all rows to be committed using wait_for_commit.
            # The predicate receives the latest committed offset token
            # (str or None) and should return True when satisfied.
            def all_rows_committed(token):
                return token is not None and int(token) >= MAX_ROWS

            channel.wait_for_commit(all_rows_committed, timeout_seconds=30)

            # Now that data has landed, check the channel status
            status = channel.get_channel_status()
            print(f"All data committed. Channel status:")
            print(f"  Committed offset:   {status.latest_committed_offset_token}")
            print(f"  Rows inserted:      {status.rows_inserted_count}")
            print(f"  Rows errored:       {status.rows_error_count}")
            print(f"  Avg server latency: {status.server_avg_processing_latency}")
            if status.rows_error_count > 0:
                print(f"  Last error:         {status.last_error_message}")

        # Channel automatically closed here
        print("Data ingestion completed")

    # Client automatically closed here


if __name__ == "__main__":
    main()

File: nodejs-example/streaming_ingest_example.js
#!/usr/bin/env node
/**
 * Example demonstrating how to use the Snowflake Streaming Ingest SDK in
 * Node.js with the high-performance architecture and default pipe.
 *
 * The default pipe is automatically created by Snowflake when you first
 * open a channel. No CREATE PIPE DDL is required. The default pipe name
 * follows the convention: <TABLE_NAME>-STREAMING
 */

"use strict";

const crypto = require("node:crypto");
const { createClient } = require("snowpipe-streaming");

const MAX_ROWS = 100_000;

// Replace these with your Snowflake object names
const DATABASE = "MY_DATABASE";
const SCHEMA = "MY_SCHEMA";
const TABLE = "MY_TABLE";

// Default pipe: Snowflake auto-creates this on first channel open
const PIPE = `${TABLE}-STREAMING`;

async function main() {
  // Create Snowflake Streaming Ingest Client
  const client = await createClient({
    clientName: `MY_CLIENT_${crypto.randomUUID()}`,
    dbName: DATABASE,
    schemaName: SCHEMA,
    pipeName: PIPE,
    profilePath: "profile.json",
  });

  console.log("Client created successfully");

  try {
    // Open a channel for data ingestion
    const { channel } = await client.openChannel({
      name: `MY_CHANNEL_${crypto.randomUUID()}`,
    });

    console.log(`Channel opened: ${channel.channelName}`);

    try {
      // Ingest rows — column names must match the target table schema.
      // The default pipe uses MATCH_BY_COLUMN_NAME to map fields.
      console.log(`Ingesting ${MAX_ROWS} rows...`);
      for (let i = 1; i <= MAX_ROWS; i++) {
        const rowId = String(i);
        await channel.appendRow(
          {
            c1: i,
            c2: rowId,
            ts: new Date(),
          },
          rowId,
        );

        if (i % 10_000 === 0) {
          console.log(`Ingested ${i} rows...`);
        }
      }

      console.log("All rows submitted. Waiting for ingestion to complete...");

      // Wait for all rows to be committed
      const expectedToken = String(MAX_ROWS);
      await channel.waitForCommit(
        (token) => token !== null && token === expectedToken,
        { timeoutMs: 60_000 },
      );

      // Verify channel status
      const status = await channel.getChannelStatus();
      console.log(
        `Latest committed offset token: ${status.latestCommittedOffsetToken}`,
      );
      console.log("All data committed successfully");
    } finally {
      await channel.close();
    }

    console.log("Data ingestion completed");
  } finally {
    await client.close();
  }
}

main().catch((err) => {
  console.error("Error during data ingestion:", err.message);
  process.exitCode = 1;
});
